### ISSUE TITLE
Question: How does app dir caching work?

### ISSUE THREAD

**Issue by khuezy** [serverless-stack/open-next#95]
Currently if you use the app dir feature of Nextjs 13, none of the "static" pages are cached.

The server-adapter has "loadHtmlPages()" that reads from `.next/server/pages-manifest.json` for `/pages`. In `appDir`, there is `.next/routes-manifest.json` that contains `staticRoutes`.

The problem w/ caching the `appDir` pages, eg: `/hello`, is that there are multiple requests to the same resource depending if it's `rsc` (prefetching). So you'd have to cache based off of the headers.
That is, the initial request to `/hello` loads the page while `<Link href="/hello" />` will prefetch the page. These are two different requests under the same URL.

How do we get CloudFront to cache based off headers? I tried adding "rsc", "next-router-prefetch|state-tree" but that breaks CF => 503.

The prefetched resources have `content-type: text/x-component`

**Comment by conico974** [comment]
The 503 part is odd, while testing for ISR i noticed this issue as well and i just added `rsc` `next-router-prefetch` `next-router-state-tree` headers to the cloudfront cache policies and it worked as expected.

It is to be noted though, that SSG pages are treated the same way as ISR ones but with a revalidate set to false, which means that they are not in pages-manifest.json as html.

The other remaining issues here, is that by default, NextServer doesn't set `cache-control` headers at all when using app dir, but it's seem to be a bug in Next.js due to this line that was missing https://github.com/vercel/next.js/blob/189e6a3687924de3172438a2912973bd49c1e37e/packages/next/src/server/base-server.ts#L2020 , but it has been resolved in a recent PR.

**Comment by khuezy** [comment]
Hmm I manually added those headers and it caused the page to hard refresh when using `<Link />` I'll have to retry to make sure I didn't do something stupid.

**Comment by khuezy** [comment]
I updated my cache policy to include rsc, next-router-prefetch, next-router-state-tree and updated the "loadHtmlPages()" to read the routes.manifest.json and now everything works as expected. My appDir pages are cached and when a dynamic data that loads on that page gets updated, it automatically by passes the cache when I reload.

@fwang what was the reason why the `rsc` stuff being removed from NextjsSite?

**Comment by khuezy** [comment]
13.3.1-canary.17 has the app dir cache in it but I don't think it did anything. I still had to read from the `routes-manifest.json` and manually set the headers for caching to work.

**Comment by conico974** [comment]
Hmm, i just updated it too, but i get the expected cache-control headers for my SSG pages `s-maxage=31536000, stale-while-revalidate`. And for my SSR pages i get `private, no-cache, no-store, max-age=0, must-revalidate`.

To be fair i've just tested it with my fix for ISR. It might be the incremental cache not working that cause some issues here.

@khuezy Do you have no cache-control at all when not using `routes-manifest.json` ?

**Comment by khuezy** [comment]
Correct, I left everything as it was and my `/app/hello/page.tsx` which simply returned a `div` was never cached. I'll double check again to make sure I didn't do anything stupid again.

**Comment by conico974** [comment]
For the SSR part i think that `no-cache` is the intended way because it's up to you to decide what should `cache-control` be on SSR request. But i don't think there is a way right now to put headers in app dir without using middleware.
The function that is supposed to set `cache-control` headers in Next is here : https://github.com/vercel/next.js/blob/8050a6c8e056efb336f15f6fa5be789cd040574d/packages/next/src/server/send-payload/revalidate-headers.ts#L5

I just tested redeploying with open-next without my fix and with next 13.3.1-canary.17 and i still have the same cache-control headers as the ones i had earlier.

**Comment by khuezy** [comment]
As @conico974 noted, you don't have to do anything in `server-adapter` to get `appDir` caching to work. You'll need to set it your `middleware`:
```
const n = NextResponse.next()
n.headers.set('cache-control', 'public, max-age=0, s-maxage=31536000, must-revalidate')
return n
```

I'll add this note in the readme in a PR later and close out this Issue. Thanks!
